### PR TITLE
CLI Framework

### PR DESCRIPTION
**Pull request by nkbt**
I reckon it is a part of #7 

This is how it works:

cli-framework

So now we have 3 "binaries":
1. `cloverfield`
2. `cf` - convenient alias for `cloverfield`
3. `cf-package` - stub for a `package` scaffold (will be implemented according to #10)

`cf` has only one command available for now: `build` with a single option `<scaffold>`. See code comments for details.

`cf build package` is equivalent to `cf-package` and starts step-by-step "wizard".
 
Regarding libs choice. I checked different commands, args and prompt libs and the most friendly seems to be `nomnom` for commands and options and `prompt` for, hm.., prompt. I used other libs before (like commander.js) and found them quite syntax-heavy and not fun.

If we are ok with tooling, I can actually implement `scaffold/package` #10 now. Should be pretty straightforward.

PS: CLI colors and messages can be customized

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#! /usr/bin/env ./node_modules/.bin/babel-node
```
Comment (nkbt) [inline index.js]: This is a nice trick to run CLI-scripts with babel-node

File: package.json [hunk 2]
```diff
@@ -15,12 +20,18 @@
     "JavaScript",
     "Node",
     "Browser",
-    "Isomorphic"
+    "Universal"
```
Comment (nkbt) [inline package.json]: :)

File: package.json [hunk 3]
```diff
@@ -15,12 +20,18 @@
...
   ],
   "author": "Eric Elliott et al...",
   "license": "MIT",
   "bugs": {
     "url": "https://github.com/ericelliott/cloverfield/issues"
   },
-  "homepage": "https://github.com/ericelliott/cloverfield"
+  "homepage": "https://github.com/ericelliott/cloverfield",
+  "dependencies": {
+    "babel": "^5.6.14",
```
Comment (nkbt) [inline package.json]: These deps are actually needed for the CLI to run, so not `devDependencies`.

File: index.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+
+import nomnom from 'nomnom';
+import glob from 'glob';
+import path from 'path';
+
+
+// Find all available scaffolds
```
Comment (nkbt) [inline index.js]: Just a bit of automation here. Not sure how valuable it is. We can always hardcode this like:
```js
import package from './scaffold/package;
const scaffolds = {
  package
};
```
But have to add new scaffolds manually here every time.

Let me know which way you prefer.

File: scaffold/package/index.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+
+const scaffold = () => {
+  prompt.get(schema, (err, {name}) =>
+    console.log(`Scaffolding ${name}...`));
+};
+
+
+// Check if script is run directly
+if (require.main === module) {
```
Comment (nkbt) [inline scaffold/package/index.js]: Another trick to let it run as independent `cf-package` binary

File: .babelrc [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
+  "stage": 1
```
Comment (nkbt) [inline .babelrc]: Needed to support `{scaffold, ...opts}` (line 25 of index.js) -  object destructuring.

File: package.json (hunk 2 above)
Comment (ericelliott) [inline package.json]: I'd leave "Isomorphic" as well until the mindshare has shifted completely to "Universal" -- that will take a while... maybe 2 years+

File: index.js (hunk 4 above)
Comment (ericelliott) [inline index.js]: Whatever you prefer is fine for now. =)

File: .babelrc (hunk 6 above)
Comment (ericelliott) [inline .babelrc]: Can you double check that this also works in `prod-module-boilerplate`? Make a new issue there if it doesn't?

File: .gitignore [hunk 7]
```diff
@@ -0,0 +1,32 @@
+# Created by .ignore support plugin (hsz.mobi)
```
Comment (ericelliott) [inline .gitignore]: If you've made general improvements to this file that would be useful to `prod-module-boilerplate`, could you make sure they're reflected there, too? I haven't looked myself...

File: index.js (hunk 1 above)
Comment (ericelliott) [inline index.js]: Cool!

File: index.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+const scaffolds = {};
+glob.sync('scaffold/*/', {realpath: true})
```
Comment (ericelliott) [inline index.js]: Add whitespace between these? =)

File: scaffold/package/index.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
+#! /usr/bin/env ./node_modules/.bin/babel-node
+
+
+import prompt from 'prompt';
+import {parse} from 'nomnom';
+
+
+const schema = {
```
Comment (ericelliott) [inline scaffold/package/index.js]: :+1:

File: scaffold/package/index.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  properties: {
+    name: {
+      description: 'Package name: ',
+      pattern: /^[a-z]+[a-z\-_]+$/,
+      message: 'Name must be only letters, numbers, dashes and underscores',
+      required: true
+    }
+  }
+};
+prompt.message = '>'.green;
```
Comment (ericelliott) [inline scaffold/package/index.js]: You could bundle these together in an object literal with `Object.assign(prompt, { <insert props here> })`...

**Comment by ericelliott** [comment]
I don't see any unit tests. Did I miss something?

**Comment by ericelliott** [comment]
Over-all, great work. :clap:

File: .gitignore (hunk 7 above)
Comment (nkbt) [inline .gitignore]: Sure

File: index.js (hunk 8 above)
Comment (nkbt) [inline index.js]: OK. Need to add this rule to ESLint http://eslint.org/docs/rules/newline-after-var

And we need eslint here as well.

File: package.json (hunk 2 above)
Comment (nkbt) [inline package.json]: Sure

File: scaffold/package/index.js (hunk 10 above)
Comment (nkbt) [inline scaffold/package/index.js]: right

**Comment by nkbt** [comment]
I wasn't sure about tests, since they require additional libs pulled into a project. As well as eslint. Was thinking of doing it in a separate PR to keep focus on CLI frameworks itself in this one.

File: index.js (hunk 8 above)
Comment (ericelliott) [inline index.js]: :+1:

**Comment by ericelliott** [comment]
Consider testing a requirement for all substantial code commits. Since we don't have the supporting stuff added here yet, feel free to add them with this initial PR. =)

**Comment by ericelliott** [comment]
And thanks for being a rockstar on this project lately. :guitar: :fire:

**Comment by nkbt** [comment]
Thanks, it's a pleasure. Always fun to do some meaningful and interesting
work.

File: .babelrc (hunk 6 above)
Comment (nkbt) [inline .babelrc]: Yes, it works fine, https://github.com/ericelliott/prod-module-boilerplate/pull/35

Feel free to discard that PR if it is not practical to keep it in there.

File: .gitignore (hunk 7 above)
Comment (nkbt) [inline .gitignore]: This one is quite common. Created if you make a new Github project and choose `Node` as default .gitignore.

Comment (nkbt) [inline .gitignore]: See https://github.com/github/gitignore/blob/master/Node.gitignore

Comment (ericelliott) [inline .gitignore]: :+1:

File: .eslintrc [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+        "key-spacing": [2, { "beforeColon": false, "afterColon": true }],
+        "quotes": [2, "single", "avoid-escape"],
+        "max-depth": [0, 4],
+        "max-len": [0, 80, 4],
+        "max-nested-callbacks": [0, 2],
+        "max-params": [0, 3],
+        "max-statements": [0, 10],
+        "new-parens": [2],
+        "new-cap": [0],
+        "newline-after-var": [2],
```
Comment (nkbt) [inline .eslintrc]: Set it to required

File: index.js (hunk 4 above)
Comment (ericelliott) [inline index.js]: In the future, we should scan for all globally installed scaffolds, but that should be a different PR. =)

File: .babelrc (hunk 6 above)
Comment (ericelliott) [inline .babelrc]: Thanks.

**Comment by nkbt** [comment]
Blue-tape does not have CLI support (the one `tape` has). I've just made a PR that fixes this issue: https://github.com/spion/blue-tape/pull/12

Until it is fixed, I have to add some automation to our tests.

File: .eslintrc (hunk 11 above)
Comment (ericelliott) [inline .eslintrc]: also in [prod-module-boilerplate](https://github.com/ericelliott/prod-module-boilerplate/edit/master/.eslintrc)? =)

Comment (nkbt) [inline .eslintrc]: Yeah, need that update there too. It actually gets bit annoying to propagate changes. I've got the same issues with my own modules. Not sure what is the best solution. Maybe some "defaults" repo with stuff and couple of scripts like `npm install && cp -f ./node_modules/defaults/.eslintrc ./`?
What do you reckon?

**Comment by ericelliott** [comment]
It does now. =)

**Comment by nkbt** [comment]
Seen it. Will remove `tape` dependency later today

**Pull request merged**